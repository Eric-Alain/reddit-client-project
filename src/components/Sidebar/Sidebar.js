//React
import React, { useEffect } from 'react'

//Redux
import { useDispatch, useSelector } from 'react-redux'
import { fetchSubreddits } from '../../state/actions/subreddits'
import { searchTyped } from '../../state/actions/search'
import { fetchData, limitDataResults } from '../../state/actions/data'

//Utils
import { unique } from '../../utils/utils'

//3rd party
import Img from 'react-cool-img'
import { Puff } from 'react-loading-icons'
import Col from 'react-bootstrap/Col'
import Row from 'react-bootstrap/Row'

const Sidebar = () => {
  //Redux
  const dispatch = useDispatch()
  const subreddits = useSelector(state => state.subreddits.children)

  //Handler
  const handleClick = str => {
    fetch(encodeURI(`https://www.reddit.com/${str}.json`))
      .then(response => response.json())
      .then(data => {
        dispatch(fetchData(data))
      })
    dispatch(searchTyped(''))
    dispatch(limitDataResults(5))
  }

  //UseEffect
  useEffect(() => {
    fetch('https://www.reddit.com/r/popular.json?geo_filter=CA&limit=30')
      .then(response => response.json())
      .then(data => {
        let subredditNameArr = unique(data.data.children.map(subreddit => subreddit.data.subreddit_name_prefixed))
        subredditNameArr.length = 20
        subredditNameArr.sort()

        let subredditDataArr = []

        subredditNameArr.forEach(name => {
          fetch(`https://www.reddit.com/${name}/about.json?limit=1`)
            .then(response => response.json())
            .then(data => {
              subredditDataArr.push(data.data)

              if (subredditDataArr.length === 20) {
                dispatch(fetchSubreddits(subredditDataArr))
              }
            })
        })
      })
  }, [dispatch])

  return (
    <Col xs={{ span: 10, order: 1 }} sm={{ span: 5, order: 2 }} md={{ span: 4, order: 2 }} xl={{ span: 3, order: 2 }} id='sidebar' className='sidebar-border bg-2 pb-3'>
      <aside>
        <Row className='mx-1 justify-content-between justify-content-md-start'>
          <Col xs='12' className='px-0'>
            <h2 className='mt-3 mb-2'>Subreddits</h2>
            <hr className='mt-0' />
          </Col>
          {/*Map data to DOM once available, limit results based on state limit*/}
          {subreddits !== undefined ? (
            subreddits.map((subreddit, i) => {
              return (
                <Col xs='6' sm='12' key={i} className='mb-3 p-1 subreddit-pill'>
                  <button className='w-100' onClick={() => handleClick(subreddit.display_name_prefixed)}>
                    <Row className='justify-content-start align-items-center'>
                      <Col xs='auto' className='ps-2 pe-0'>
                        <Img
                          className='subreddit-img'
                          src={
                            subreddit.icon_img !== '' && subreddit.icon_img !== null && subreddit.icon_img !== undefined
                              ? subreddit.icon_img
                              : 'https://logodownload.org/wp-content/uploads/2018/02/reddit-logo-16.png'
                          }
                          alt=''
                        />
                      </Col>
                      <Col xs='auto' className='ps-2 ps-md-auto'>
                        <small>{subreddit.display_name}</small>
                      </Col>
                    </Row>
                  </button>
                </Col>
              )
            })
          ) : (
            /*If data unavailable, insert a placeholder element*/
            <Puff className='loading' />
          )}
        </Row>
      </aside>
    </Col>
  )
}

export default Sidebar
