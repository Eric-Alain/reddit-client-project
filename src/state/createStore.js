//Redux
import { createStore as reduxCreateStore, combineReducers } from 'redux'
import { composeWithDevTools } from 'redux-devtools-extension'

//Reducers
import searchReducer from '../state/reducers/search'
import dataReducer from '../state/reducers/data'
import subredditsReducer from '../state/reducers/subreddits'
import togglesReducer from '../state/reducers/toggles'

//Create root reducer
export const rootReducer = combineReducers({
  search: searchReducer,
  data: dataReducer,
  subreddits: subredditsReducer,
  toggles: togglesReducer
})

//Create store, pass preloadedState via plugin
const createStore = preloadedState => {
  return reduxCreateStore(rootReducer, preloadedState, composeWithDevTools())
}
export default createStore
