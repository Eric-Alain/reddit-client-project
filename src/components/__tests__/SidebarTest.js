//React
import React from 'react'

//Redux
import { Provider } from 'react-redux'
import createStore from '../../state/createStore'

//Components
import Sidebar from '../Sidebar/Sidebar'

//Testing
import renderer from 'react-test-renderer'

//3rd party
import registerIcons from '../../icons/icons'

const store = createStore()
registerIcons()

describe('Sidebar', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Sidebar />
        </Provider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
