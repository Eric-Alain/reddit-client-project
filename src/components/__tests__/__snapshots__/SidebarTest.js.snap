// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar renders correctly 1`] = `
<div
  className="sidebar-border bg-2 pb-3 col-xl-3 col-md-4 col-sm-5 col-10 order-xl-2 order-md-2 order-sm-2 order-1"
  id="sidebar"
>
  <aside>
    <div
      className="mx-1 justify-content-between justify-content-md-start row"
    >
      <div
        className="px-0 col-12"
      >
        <h2
          className="mt-3 mb-2"
        >
          Subreddits
        </h2>
        <hr
          className="mt-0"
        />
      </div>
      <svg
        className="icon-loading loading"
        height={44}
        viewBox="0 0 44 44"
        width={44}
      >
        <g
          fill="none"
          fillRule="evenodd"
          stroke="#fff"
          strokeWidth={2}
        >
          <circle
            cx={22}
            cy={22}
            r={1}
          >
            <animate
              attributeName="r"
              begin="0s"
              calcMode="spline"
              dur="1.8s"
              keySplines="0.165, 0.84, 0.44, 1"
              keyTimes="0; 1"
              repeatCount="indefinite"
              values="1; 20"
            />
            <animate
              attributeName="stroke-opacity"
              begin="0s"
              calcMode="spline"
              dur="1.8s"
              keySplines="0.3, 0.61, 0.355, 1"
              keyTimes="0; 1"
              repeatCount="indefinite"
              values="1; 0"
            />
          </circle>
          <circle
            cx={22}
            cy={22}
            r={1}
          >
            <animate
              attributeName="r"
              begin="-0.9s"
              calcMode="spline"
              dur="1.8s"
              keySplines="0.165, 0.84, 0.44, 1"
              keyTimes="0; 1"
              repeatCount="indefinite"
              values="1; 20"
            />
            <animate
              attributeName="stroke-opacity"
              begin="-0.9s"
              calcMode="spline"
              dur="1.8s"
              keySplines="0.3, 0.61, 0.355, 1"
              keyTimes="0; 1"
              repeatCount="indefinite"
              values="1; 0"
            />
          </circle>
        </g>
      </svg>
    </div>
  </aside>
</div>
`;
